app/workers/backend_delete_application_worker.rb

```diff
@@ -0,0 +1,38 @@
+# frozen_string_literal: true
+
+class BackendDeleteApplicationWorker < ApplicationJob
+  queue_as :low
+
+  def perform(event_id)
+    event = EventStore::Repository.find_event!(event_id)
+
+    @application = build_application(event)
+
+    [ApplicationKey, ReferrerFilter].each(&method(:delete_backend_associations))
+
+    application.delete_backend_application
+  rescue ActiveRecord::RecordNotFound => exception
+    System::ErrorReporting.report_error(exception, parameters: {event_id: event_id})
+  end
+
+  private
+
+  attr_reader :application
+
+  def build_application(event)
+    app_attrs = {id: event.application.id, application_id: event.application_id}
+    Cinstance.new(app_attrs, without_protection: true).tap do |app|
+      app.service = Service.new({id: event.service_id}, without_protection: true)
+    end
+  end
+
+  def build_association_object(deleted_object)
+    deleted_object.object_instance { |object| object.application = application }
+  end
+
+  def delete_backend_associations(klass)
+    DeletedObject.public_send(klass.to_s.underscore.pluralize).where(owner: application).order(:id).find_each do |deleted_object|
```

Comment:
This works now because we know it's a `has_many`, not a `has_one`. Do you think we it's worth extracting `klass.to_s.underscore.pluralize` to a separate method? Ideally, we would resolve the association name from the name of the class without assuming much, but I think it may already help a lot whoever reads the code realise the importance of this and, in case of a `has_one` association of `Cinstance` added later to line 11, where to fix it. Plus, you could alternatively map the list of classes into association names (using this new method) and be more assertive with `#delete_backend_associations` by passing the name of the association instead of the name of the class.

Reply:
@guicassolato this is removed in the following PR 😉

Reply:
I mean #1832